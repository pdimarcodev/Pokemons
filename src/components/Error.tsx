import { StyleSheet, View, Text, Button } from "react-native";

interface Props {
  retry: () => void;
}

export default function ErrorMessage({ retry }: Props) {
  return (
    <View style={styles.container}>
      <Text style={styles.text}>An error occurred.</Text>
      <Button title="Retry" onPress={retry} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  text: {
    color: "red",
    fontSize: 20,
  },
});
