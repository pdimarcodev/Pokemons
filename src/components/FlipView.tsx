import { ReactNode } from "react";
import { StyleProp, StyleSheet } from "react-native";
import Animated, {
  Easing,
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";
import {
  GestureHandlerRootView,
  HandlerStateChangeEvent,
  State,
  TapGestureHandler,
  TapGestureHandlerEventPayload,
} from "react-native-gesture-handler";

interface Props {
  containerStyle: StyleProp<Object>;
  RegularContent: ReactNode;
  FlippedContent: ReactNode;
}

export default function FlipCard({
  containerStyle,
  RegularContent,
  FlippedContent,
}: Props) {
  const rotation = useSharedValue(0);

  const toggleFlip = () => {
    rotation.value = withTiming(rotation.value + 180, {
      duration: 500,
      easing: Easing.ease,
    });
  };

  const onTap = ({
    nativeEvent,
  }: HandlerStateChangeEvent<TapGestureHandlerEventPayload>) => {
    if (nativeEvent.state === State.END) {
      toggleFlip();
    }
  };

  const frontCardStyle = useAnimatedStyle(() => {
    const interpolatedRotateY = interpolate(
      rotation.value,
      [0, 180, 360],
      [0, 180, 0]
    );
    return {
      transform: [
        { perspective: 1000 },
        { rotateY: `${interpolatedRotateY}deg` },
      ],
    };
  });

  const backCardStyle = useAnimatedStyle(() => {
    const interpolatedRotateY = interpolate(
      rotation.value,
      [0, 180, 360],
      [180, 360, 180]
    );
    return {
      transform: [
        { perspective: 1000 },
        { rotateY: `${interpolatedRotateY}deg` },
      ],
    };
  });

  return (
    <GestureHandlerRootView>
      <TapGestureHandler onHandlerStateChange={onTap}>
        <Animated.View
          style={[containerStyle, frontCardStyle, styles.cardContainer]}
        >
          {RegularContent}
        </Animated.View>
      </TapGestureHandler>
      <TapGestureHandler onHandlerStateChange={onTap}>
        <Animated.View
          style={[
            containerStyle,
            backCardStyle,
            styles.cardContainer,
            styles.cardBack,
          ]}
        >
          {FlippedContent}
        </Animated.View>
      </TapGestureHandler>
    </GestureHandlerRootView>
  );
}

const styles = StyleSheet.create({
  cardContainer: {
    backfaceVisibility: "hidden",
  },
  cardBack: {
    position: "absolute",
  },
});
