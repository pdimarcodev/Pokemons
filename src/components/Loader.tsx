import {
  ActivityIndicator,
  ActivityIndicatorProps,
  StyleSheet,
  View,
} from "react-native";

interface Props extends ActivityIndicatorProps {}

export default function Loader(props: Props) {
  return (
    <View style={styles.loader}>
      <ActivityIndicator {...props} />
    </View>
  );
}

const styles = StyleSheet.create({
  loader: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#111111",
  },
});
