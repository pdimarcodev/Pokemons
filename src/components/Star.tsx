import Animated, { StyleProps, ZoomIn, ZoomOut } from "react-native-reanimated";
import { AntDesign } from "@expo/vector-icons";

interface Props {
  size: number;
  style?: StyleProps;
}

export function Star({ size, style }: Props) {
  return (
    <Animated.View entering={ZoomIn} exiting={ZoomOut} style={style}>
      <AntDesign name="star" size={size} color="#ff4d6d" />
    </Animated.View>
  );
}
