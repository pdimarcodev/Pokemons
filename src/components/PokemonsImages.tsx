import { memo } from "react";
import { Image, StyleSheet } from "react-native";
import FlipView from "./FlipView";
import { Sprites } from "@/interfaces";

type ImageType = "default" | "shiny" | "female" | "shiny_female";

interface PokemonsImagesProps {
  imageType: ImageType;
  sprites?: Sprites;
}

interface PokemonImageProps {
  uri: string;
}

const PokemonImage = memo(({ uri }: PokemonImageProps) => (
  <Image source={{ uri }} style={styles.image} resizeMode="stretch" />
));

export default function PokemonsImages({
  imageType,
  sprites,
}: PokemonsImagesProps) {
  if (!sprites) return null;

  return (
    <FlipView
      containerStyle={styles.cardContainer}
      RegularContent={
        <PokemonImage uri={sprites[`front_${imageType}`] as string} />
      }
      FlippedContent={
        <PokemonImage uri={sprites[`back_${imageType}`] as string} />
      }
    />
  );
}

const styles = StyleSheet.create({
  cardContainer: {
    height: 150,
    width: 150,
    marginVertical: 10,
  },
  image: {
    width: "100%",
    aspectRatio: 1,
  },
});
