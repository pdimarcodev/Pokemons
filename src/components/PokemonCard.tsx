import { useMemo } from "react";
import { Image, Pressable, StyleSheet, Text, View } from "react-native";
import { useRouter } from "expo-router";
import Animated, {
  Easing,
  FadeInUp,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSequence,
  withSpring,
  withTiming,
} from "react-native-reanimated";
import Loader from "./Loader";
import { useAppContext } from "@/context/AppContext";
import { useGetPokemonByName } from "@/hooks/useGetPokemonByName";
import { useFavorites } from "@/hooks/useFavorites";
import { capitalize } from "@/utils/capitalize";

interface Props {
  name: string;
  index: number;
}

const DELAY_FACTOR = 100;
const PressableAnimated = Animated.createAnimatedComponent(Pressable);

export const PokemonCard = ({ name, index }: Props) => {
  const { data, error, isValidating } = useGetPokemonByName({ name });
  const { favorites } = useAppContext();
  const { addFavorite, removeFavorite } = useFavorites();
  const router = useRouter();
  const scale = useSharedValue(1);
  const formattedPokemonsName = useMemo(() => capitalize(name), [name]);

  const animatedContainerStyle = useAnimatedStyle(() => {
    return {
      transform: [{ scale: scale.value }],
    };
  });

  function goToDetails() {
    router.navigate({
      pathname: `/pokemons/${name}`,
      params: { formattedPokemonsName },
    });
  }

  function onPress() {
    scale.value = withSequence(
      withTiming(1.1, { easing: Easing.bounce, duration: 500 }),
      withSpring(1, undefined, (isFinished) => {
        if (isFinished) {
          runOnJS(goToDetails)();
        }
      })
    );
  }

  function onLongPress() {
    // console.warn("Add to favorites");
    if (favorites?.includes(name)) {
      removeFavorite(name);
    } else {
      addFavorite(name);
    }
  }

  return (
    <PressableAnimated
      onPress={onPress}
      onLongPress={onLongPress}
      style={[styles.card, animatedContainerStyle]}
    >
      {isValidating || typeof favorites === undefined ? (
        <Loader size="small" />
      ) : (
        <Animated.View
          entering={FadeInUp.delay(DELAY_FACTOR * index).springify()}
        >
          <View style={styles.imageContainer}>
            <Image
              source={{ uri: data?.sprites?.front_shiny }}
              style={styles.image}
              resizeMode="stretch"
            />
          </View>
          <View style={styles.textContainer}>
            <Text numberOfLines={1} adjustsFontSizeToFit style={styles.name}>
              {formattedPokemonsName}
            </Text>
            <Text style={styles.id}>#{`${data?.id}`}</Text>
          </View>
        </Animated.View>
      )}
    </PressableAnimated>
  );
};

const styles = StyleSheet.create({
  card: {
    flex: 1,
    height: 130,
    borderRadius: 5,
    borderColor: "#0077B6",
    borderWidth: 1,
    borderBottomWidth: 2,
    paddingBottom: 10,
    backgroundColor: "#003566",
  },
  imageContainer: {
    width: 90,
    height: 90,
    alignSelf: "center",
  },
  image: {
    flex: 1,
  },
  textContainer: {
    flexDirection: "row",
    alignItems: "baseline",
    justifyContent: "space-between",
    paddingHorizontal: 10,
  },
  name: {
    flexShrink: 1,
    color: "#fff",
    fontWeight: "bold",
    fontSize: 15,
    marginRight: 2,
  },
  id: {
    color: "#fff",
    fontSize: 11,
  },
  loader: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
});
