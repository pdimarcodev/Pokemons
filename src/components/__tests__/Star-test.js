import React from "react";
import renderer from "react-test-renderer";
import { Star } from "@/components/Star";

jest.mock("@expo/vector-icons/AntDesign", () =>
  jest.requireActual("react-native-vector-icons/AntDesign")
);

describe("Star Component", () => {
  it("renders the favorite star", async () => {
    const tree = renderer.create(<Star isFavorite size={24} />);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(tree.toJSON()).toMatchSnapshot();

    const icon = tree.root.findByType("Text");
    expect(icon.props.style[0].fontSize).toEqual(24);
  });

  it("renders the non-favorite star", async () => {
    const tree = renderer.create(<Star isFavorite={false} size={50} />);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(tree.toJSON()).toMatchSnapshot();

    const icon = tree.root.findByType("Text");
    expect(icon.props.style[0].fontSize).toEqual(50);
  });
});
