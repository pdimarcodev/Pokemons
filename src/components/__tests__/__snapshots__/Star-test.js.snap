// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Star Component renders the favorite star 1`] = `
<View
  collapsable={false}
  entering={[Function]}
  exiting={[Function]}
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#ff4d6d",
          "fontSize": 24,
        },
        undefined,
        {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Star Component renders the non-favorite star 1`] = `
<View
  collapsable={false}
  entering={[Function]}
  exiting={[Function]}
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#ff4d6d",
          "fontSize": 50,
        },
        undefined,
        {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
</View>
`;
